pkg/clustermanager/kube_proxy.go

```diff
@@ -0,0 +1,505 @@
+	return u.retrier.Retry(func() error {
+		return up.PrepareForUpgrade(ctx, u.log, managementClusterClient, workloadClusterClient, spec)
+	})
+}
+
+// CleanupAfterUpgrade perfoms the necessary steps after an upgrade.
+func (u KubeProxyCLIUpgrader) CleanupAfterUpgrade(ctx context.Context,
+	spec *cluster.Spec,
+	managementClusterKubeconfigPath, workloadClusterKubeconfigPath string,
+) error {
+	managementClusterClient, workloadClusterClient, err := u.buildClients(
+		managementClusterKubeconfigPath, workloadClusterKubeconfigPath,
+	)
+	if err != nil {
+		return err
+	}
+
+	up := NewKubeProxyUpgrader()
+
+	return u.retrier.Retry(func() error {
+		return up.CleanupAfterUpgrade(ctx, u.log, managementClusterClient, workloadClusterClient, spec)
+	})
+}
+
+func (u KubeProxyCLIUpgrader) buildClients(
+	managementClusterKubeconfigPath, workloadClusterKubeconfigPath string,
+) (managementClusterClient, workloadClusterClient client.Client, err error) {
+	u.log.V(4).Info("Building client for management cluster", "kubeconfig", managementClusterKubeconfigPath)
+	managementClusterClient, err = u.clientFactory.BuildClientFromKubeconfig(managementClusterKubeconfigPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	u.log.V(4).Info("Building client for workload cluster", "kubeconfig", workloadClusterKubeconfigPath)
+	workloadClusterClient, err = u.clientFactory.BuildClientFromKubeconfig(workloadClusterKubeconfigPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return managementClusterClient, workloadClusterClient, nil
+}
+
+// NewKubeProxyUpgrader builds a new KubeProxyUpgrader.
+func NewKubeProxyUpgrader(opts ...KubeProxyUpgraderOpt) KubeProxyUpgrader {
+	u := &KubeProxyUpgrader{
+		updateKubeProxyRetries: 30,
+		updateKubeProxyBackoff: 2 * time.Second,
+	}
+
+	for _, opt := range opts {
+		opt(u)
+	}
+
+	return *u
+}
+
+// KubeProxyUpgrader prepares a cluster for a kube-proxy upgrade.
+type KubeProxyUpgrader struct {
+	updateKubeProxyRetries int
+	updateKubeProxyBackoff time.Duration
+}
+
+// KubeProxyUpgraderOpt allows to customize a KubeProxyUpgraderOpt
+// on construction.
+type KubeProxyUpgraderOpt func(*KubeProxyUpgrader)
+
+// WithUpdateKubeProxyTiming allows to customize the retry paramenter for the
+// kube-proxy version update. This is for unit tests.
+func WithUpdateKubeProxyTiming(retries int, backoff time.Duration) KubeProxyUpgraderOpt {
+	return func(u *KubeProxyUpgrader) {
+		u.updateKubeProxyRetries = retries
+		u.updateKubeProxyBackoff = backoff
+	}
+}
+
+// PrepareForUpgrade gets the workload cluster ready for a smooth transition between the
+// old kube-proxy that always uses iptables legacy and the new one that detects the host preference
+// and is able to work with nft as well. This is idempotent, so it can be called in a loop if transient
+// errors are a risk.
+func (u KubeProxyUpgrader) PrepareForUpgrade(ctx context.Context, log logr.Logger, managementClusterClient, workloadClusterClient client.Client, spec *cluster.Spec) error {
```

Comment:
does this handle case like upgrading from one eks-a version to another one with node count scaling up?

Reply:
I haven't tested that specific scenario but yes. It doesn't matter if the number of new nodes is the same as the old nodes or not: all old nodes will use the "legacy" kube-proxy and all new nodes will use the "new" kube-proxy version from the original DS.